src/alloc_bytes.rs

```diff
@@ -0,0 +1,113 @@
+use std::alloc;
+use std::alloc::Layout;
+use std::slice;
+use std::borrow::Cow;
+
+use rustc_target::abi::{Align, Size};
+use rustc_middle::mir::interpret::AllocBytes;
+
+/// Allocation bytes that explicitly handle the layout of the data they're storing.
+/// This is necessary to interface with native code that accesses the program store in Miri.
+#[derive(Debug)]
+#[derive(PartialEq, Eq, Hash)] // TODO Necessary?
+pub struct MiriAllocBytes {
+    /// Stored layout information about the allocation.
+    layout: alloc::Layout,
+    /// Pointer to the first byte of the allocation.
+    /// Invariant: if `self.layout.size() == 0`
+    /// * then `self.ptr == self.layout.align()`
+    /// * else `self.ptr` points to memory allocated with `self.layout`. 
```

Comment:
```suggestion
    /// Invariant: 
    /// * If `self.layout.size() == 0`, then `self.ptr` is some suitably aligned pointer
    ///   without provenance (and no actual memory was allocated).
    /// * Otherwise, `self.ptr` points to memory allocated with `self.layout`. 
```